File: src/main/java/net/tecdroid/subsystems/drivetrain/SwerveDriveDriver.java
package net.tecdroid.subsystems.drivetrain;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.LinearVelocity;

import java.util.function.Supplier;

import static edu.wpi.first.units.Units.MetersPerSecond;
import static edu.wpi.first.units.Units.RadiansPerSecond;

public class SwerveDriveDriver {

    Supplier<LinearVelocity>  forwardsVelocitySupplier;
    Supplier<LinearVelocity>  sidewaysVelocitySupplier;
    Supplier<AngularVelocity> angularVelocitySupplier;
    DriveOrientation orientation = DriveOrientation.FIELD_ORIENTED;

    public SwerveDriveDriver() {
        this.forwardsVelocitySupplier = MetersPerSecond::zero;
        this.sidewaysVelocitySupplier = MetersPerSecond::zero;
        this.angularVelocitySupplier = RadiansPerSecond::zero;
    }

    public ChassisSpeeds obtainTargetSpeeds(Rotation2d currentAngle) {
        LinearVelocity  vx = forwardsVelocitySupplier.get();
        LinearVelocity  vy = sidewaysVelocitySupplier.get();
        AngularVelocity w  = angularVelocitySupplier.get();

        return isFieldOriented() ? ChassisSpeeds.fromFieldRelativeSpeeds(vx, vy, w, currentAngle) :
            new ChassisSpeeds(vx, vy, w);
    }

    public void toggleOrientation() {
        orientation = switch (orientation) {
            case FIELD_ORIENTED -> DriveOrientation.ROBOT_ORIENTED;
            case ROBOT_ORIENTED -> DriveOrientation.FIELD_ORIENTED;
        };
    }

    private boolean isFieldOriented() {
        return orientation == DriveOrientation.FIELD_ORIENTED;
    }

    public void setLongitudinalVelocitySupplier(Supplier<LinearVelocity> forwardsVelocitySupplier) {
        this.forwardsVelocitySupplier = forwardsVelocitySupplier;
    }

    public void setTransversalVelocitySupplier(Supplier<LinearVelocity> sidewaysVelocitySupplier) {
        this.sidewaysVelocitySupplier = sidewaysVelocitySupplier;
    }

    public void setAngularVelocitySupplier(Supplier<AngularVelocity> angularVelocitySupplier) {
        this.angularVelocitySupplier = angularVelocitySupplier;
    }

    public enum DriveOrientation {
        FIELD_ORIENTED,
        ROBOT_ORIENTED
    }
}

File: src/main/java/net/tecdroid/subsystems/drivetrain/SwerveDriveDriver.kt
package net.tecdroid.subsystems.drivetrain

import edu.wpi.first.math.filter.SlewRateLimiter
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.units.Units.Hertz
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.AngularVelocity
import edu.wpi.first.units.measure.LinearVelocity
import edu.wpi.first.units.measure.Time
import edu.wpi.first.util.sendable.Sendable
import edu.wpi.first.util.sendable.SendableBuilder
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.Commands
import net.tecdroid.constants.integratorTabName
import net.tecdroid.util.units.toRotation2d

class SwerveDriveDriver(
    private val maxLinearVelocity: LinearVelocity,
    private val maxAngularVelocity: AngularVelocity,
    accelerationPeriod: Time,
    decelerationPeriod: Time = accelerationPeriod
) {
    var longitudinalVelocityFactorSource = { 0.0 }
    var transversalVelocityFactorSource = { 0.0 }
    var angularVelocityFactorSource = { 0.0 }

    private val da = accelerationPeriod.asFrequency()
    private val dd = -decelerationPeriod.asFrequency()

    private val longitudinalRateLimiter = SlewRateLimiter(da.`in`(Hertz), dd.`in`(Hertz), 0.0)
    private val transversalRateLimiter = SlewRateLimiter(da.`in`(Hertz), dd.`in`(Hertz), 0.0)
    private val angularRateLimiter = SlewRateLimiter(da.`in`(Hertz), dd.`in`(Hertz), 0.0)

    private var orientation = DriveOrientation.FieldOriented

    private val isFieldOriented: Boolean
        get() = orientation == DriveOrientation.FieldOriented

    fun obtainTargetSpeeds(currentAngle: Angle): ChassisSpeeds {
        val xf = longitudinalVelocityFactorSource()
        val yf = transversalVelocityFactorSource()
        val wf = angularVelocityFactorSource()

        val vx = maxLinearVelocity * longitudinalRateLimiter.calculate(xf)
        val vy = maxLinearVelocity * transversalRateLimiter.calculate(yf)
        val vw = maxAngularVelocity * angularRateLimiter.calculate(wf)

        return if (isFieldOriented) ChassisSpeeds.fromFieldRelativeSpeeds(
            vx,
            vy,
            vw,
            currentAngle.toRotation2d()
        ) else ChassisSpeeds(vx, vy, vw)
    }

    fun toggleOrientation() {
        orientation = when (orientation) {
            DriveOrientation.FieldOriented -> DriveOrientation.RobotOriented
            DriveOrientation.RobotOriented -> DriveOrientation.FieldOriented
        }
    }

    fun toggleOrientationCommand() : Command = Commands.runOnce(::toggleOrientation)

    fun setFieldOriented() {
        orientation = DriveOrientation.FieldOriented
    }

    fun setFieldOrientedCommand() : Command = Commands.runOnce(::setFieldOriented)

    fun setRobotOriented() {
        orientation = DriveOrientation.RobotOriented
    }

    fun setRobotOrientedCommand() : Command = Commands.runOnce(::setRobotOriented)

    enum class DriveOrientation {
        FieldOriented,
        RobotOriented
    }
}